### ISSUE TITLE
B2C Token Caching for multiple resources not working

### ISSUE THREAD

**Issue by koesclem** [AzureAD/microsoft-authentication-library-for-android#904]
Hi,

I'm trying to use the Android MSAL with a B2C authority. Basically I have to fetch multiple Access Token for multiple resources. As I understand requesting multiple scopes for multiple resources at once is not possible, so I have to fetch my token multiple times interactively. So here is my issue/possible library bug which I will try to describe in as much detail as possible:

When fetching token interactively or silently, the MSAL always adds the default scopes "openid offline_access profile". This seems fine to me - but now if a valid access token is returned the library is caching/saving this token via `MsalOAuth2TokenCache`'s` save(oAuth2Strategy request response)` method. Although the "openid offline_access profile" scopes are not part of the raw access token response returned, the `AccessTokenRecord` to save is aggregated with those default scopes before saving - see `MicrosoftStsAccountCredentialAdapter`'s `createAccessToken` method. I have found following comment in the code regarding this behaviour:

``` ===============================================================
            NOTE: When requesting tokens for resources other than MS Graph or AAD Graph the default scopes
            openid, profile and offline_access are not returned.  They need to be written into the cache anyway
            to avoid cache misses.
            ===============================================================
```

So I can somehow see why this was added but now here is the thing: When the library is performing `MsalOAuth2TokenCache`'s `saveCredentials(final Credential... credentials)` method later, it also checks if scopes of Access Token in the cache are intersecting and deleting those token if they do. But since every token saved in the cache always has "openid offline_access profile" scopes automatically added they **always** intersect.

So ihmo this causes an issue when trying to fetch multiple token for different scopes/resources since the library is not able to hold multiple token at the same time effectively preventing one to being able to fetch them later silently from the cache. The iOS Version of the MSAL does not seem to have this problem since token cache records are not accumulated with mentioned default scopes and related intersection checks work as intended.

As far as I can see caching multiple token at the same time did not ever work in the Android MSAL for this case. I'm happy to provide any further information if needed and would be glad to receive feedback on this issue.

**Comment by iambmelt** [comment]
@koesclem After you perform an interactive auth (using `acquireToken()`) for a given B2C policy you'll have a `refresh_token` in your cache that can be used for subsequent `acquireTokenSilent()` calls against that policy. Additional _interactive_ calls against that policy shouldn't be necessary unless that policy requires interaction to function (such as a `change_password` policy, policy requiring MFA challenge, or if your refresh token expires due to revocation or age).

You're correct that MSAL on Android doesn't cache multiple `access_tokens` for a given client/resource; this is by-design as the `refresh_token` can be used to silently request `access_tokens` of varying scopes for a given user so long as the user (or admin) has previously granted consent.

As it turns out, requesting scopes spanning multiple resources _is_ allowed as part of the consent model (see [`withOtherScopesToAuthorize()`](https://github.com/AzureAD/microsoft-authentication-library-for-android/blob/dce3e35353c19f7d821d6c0896e570d0fb503a5e/msal/src/main/java/com/microsoft/identity/client/AcquireTokenParameters.java#L163-L166)) -- the additional scopes will _not_ be signed into the `access_token` your client receives from AAD, however the additional scopes will be pre-consented by the user meaning that subsequent calls to `acquireTokenSilent` will be able to use the `refresh_token` to fetch an `access_token` which contains those granted scopes

In case you've not seen it, I'll provide a link over to our B2C sample:
https://github.com/Azure-Samples/ms-identity-android-java#b2cmodefragment-class

My suggestion would be to try and use `withOtherScopesToAuthorize` to get the user to pre-consent and then use `acquireTokenSilent` to fetch tokens for those resources when your app needs them. As a friendly hint, remember that B2C treats each policy as a _separate authority_ meaning that the `access_tokens`, `refresh_tokens` and `id_tokens` returned by each policy cannot be used interchangeably (AKA the `refresh_token` for Policy A cannot be used to request an `access_token` for Policy B). Additionally, each policy will return a separate `IAccount` object so when providing the account to `acquireTokenSilent()` with `forAccount()` the same `IAccount` which was used to execute the interactive policy must be used to execute the silent policy.

Hope this helps, feel free to provide logs if none of the above unblocks and you believe you're still seeing misbehavior or a broken scenario

**Comment by koesclem** [comment]
@iambmelt Thank you for your immediate reply - I fully agree with you on your first paragraph as this is also my understanding of the process. I can also agree to the fact that MSAL should only cache a single token for a given resource.

But as mentioned in my case, I have to fetch Token for **multiple** resources.  Since AAD `access_tokens` can't have multiple audiences, multipe `access_tokens` are required to authorize for multiple resources (please correct me if I am mistaken here).

As you suggested, I have already tried to request consent for additional scopes of other resources during initial sign-in by using `withOtherScopesToAuthorize()`. This ends up with the same error message I would get when not using this parameter (by providing the scopes via `withScopes()`):

```
The scope 'openid offline_access profile scope_of_resource1 scope_of_resource2'
provided in request specifies more than one resource for an access token, which is not supported.
```

I also checked the resulting url which is opened in the browser when using `withOtherScopesToAuthorize()`. The url is not different for when setting all scopes via `withScopes()`. To be clear: I'm setting `scope_of_resource1` via `withScopes()` and `scope_of_resource2` via `withOtherScopesToAuthorize()` which results in the mentioned error. So `withOtherScopesToAuthorize()` does not seem to have any impact. The url query parameter set is: 

`&scope=openid%20offline_access%20profile%20scope_of_resource1%20scope_of_resource2`

I would also like to add that I am only using a single B2C policy.

Another more unrelated note would be, that I initially did not know about the https://github.com/Azure-Samples/ms-identity-android-java#b2cmodefragment-class sample class, since I used the kotlin sample as a reference (https://github.com/Azure-Samples/ms-identity-android-kotlin) which apparently does not contain the same feature set as the the java sample.

Thank you again for your feedback and I hope you are able to help on this ongoing issue.

**Comment by iambmelt** [comment]
Linking these two issues, as they appear to be the same:
https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/550

**Comment by iambmelt** [comment]
@koesclem Reviewing the [above linked thread](https://github.com/AzureAD/microsoft-authentication-library-for-android/issues/904#issuecomment-578285712), I suspect this may be service limitation where the B2C STS exhibits different behavior than AAD.

I'm going to reach out to my PM/B2C contacts to figure out what the support story is here. FWIW this contributor [experimented with some workarounds](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/550#issuecomment-388325315) using out DotNet library. I haven't independently verified this so I can't speak to it but you may try setting `Prompt.Consent` for these resources individually and seeing if you can minimally skip the prompt for a password

**Comment by koesclem** [comment]
@iambmelt Thank you for reaching out to your contacts. I am familiar with your mentioned [thread](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/550) and agree with the opinion voiced there, that opening the browser multiple times (also when it is just flashing up shortly) is no feasible solution in a mobile context.

The [workaround](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/550#issuecomment-388325315) of fetching tokens interactively a few times with subsequent silent requests would lead us back to my initial post on this issue. Namely not being able to save the results to the cache (and later perform lookups) since the library-set default scopes intersect. I just wanted to re-mention this to emphasize how interlinked those topics are.

I am excited to hear what you can find out regarding this.

**Comment by jmprieur** [comment]
@koesclem @iambmelt I believe that the scopes would accumulate in the access token when you call AcquireTokenSilent. You should be fine ?

**Comment by koesclem** [comment]
@jmprieur @iambmelt As far as I know having scopes of different resources in AAD B2C `access_token` is not supported. The `access_token`'s audience claim does only identify a **single** intended recipient of the `access_token`. Even if scopes would accumulate (which they don't) this would not be of any help since the audience claim only supporting a single audience is, as I understand it, security related and by design.
Of course I also checked the `access_token`'s payload to confirm this.

To further add information to this issue I can shortly describe how interactively acquiring the token for `scope_of_resource1` and later silently acquiring the token for `scope_of_resource2` fails:

First I would like to mention that fetching `access_token` interactively is working for both `scope_of_resource1` & `scope_of_resource2` independently - with the constraint being that they overwrite each other in the cache (see my initial [Post](https://github.com/AzureAD/microsoft-authentication-library-for-android/issues/904#issue-554643060)).

When requesting the `scope_of_resource2` later, MSAL throws an `MsalClientException` wrapping a `NullPointerException` since no `access_token` is returned unexpectedly.
When taking a closer look at the AcquireTokenSilent Flow, MSAL is initially (and imo correctly) not able to find a valid `access_token` in the retrieved `ICacheRecord` for the new requested `scope_of_resource2` - see `LocalMSALController`

```
final ICacheRecord fullCacheRecord = cacheRecords.get(0);

if (accessTokenIsNull(fullCacheRecord)
                || refreshTokenIsNull(fullCacheRecord)
                || parameters.getForceRefresh()
                || !isRequestAuthorityRealmSameAsATRealm(parameters.getAuthority(), fullCacheRecord.getAccessToken())) {
            if (!refreshTokenIsNull(fullCacheRecord)) {
                // No AT found, but the RT checks out, so we'll use it
                Logger.verbose(
                        TAG + methodName,
                        "No access token found, but RT is available."
                );

                renewAccessToken(
...
```

At the same time MSAL deems the cached `refresh_token` as being able to fetch the `access_token` for the new scope of the new resource `scope_of_resource2` by calling the `renewAccessToken()` method. This is because the `MsalOAuth2TokenCache` regards the `AccountRecord`'s `authorityType` as being "multiResourceCapable"

```
final boolean isMultiResourceCapable = MicrosoftAccount.AUTHORITY_TYPE_V1_V2.equals(
                account.getAuthorityType()
        );
```

Sadly I was not able to find any documentation to this "multi-Resource-Capability". As mentioned, the AcquireTokenSilent Flow finally results in an error while performing the token renewal.

I hope this gives further insights to eventually find a solution for this issue.

**Comment by iambmelt** [comment]
> When requesting the `scope_of_resource2` later, MSAL throws an `MsalClientException` wrapping a `NullPointerException` since no `access_token` is returned unexpectedly.

Can you include this stack trace for this error? This error shouldn't be bubbling up like this...

**Comment by iambmelt** [comment]
@koesclem Just providing an update here, still working with PM contacts to capture expected behavior here from both SDK and B2C backend. Once I have the prescribed guidance I'll share here and spec-out any changes to MSAL as necessary. Any PRs for hotfixes developed will be shared into this issue for you to see

So far, this seems to be the convergence of two issues:
- Differing service behavior between B2C and AAD as concerns token issuance when multiple resources are involved
- Differing cache persistence/lookup strategies across MSAL platforms (iOS v Android)

Thanks for your patience

**Comment by koesclem** [comment]
> Can you include this stack trace for this error? This error shouldn't be bubbling up like this...

@iambmelt Yes, of course - here is the stacktrace of the mentioned `MsalClientException`:
```
com.microsoft.identity.client.exception.MsalClientException: Attempt to invoke virtual method 'long java.lang.Long.longValue()' on a null object reference
	at com.microsoft.identity.client.internal.controllers.MsalExceptionAdapter.msalExceptionFromBaseException(MsalExceptionAdapter.java:51)
	at com.microsoft.identity.client.SingleAccountPublicClientApplication$3.onError(SingleAccountPublicClientApplication.java:242)
	at com.microsoft.identity.client.SingleAccountPublicClientApplication$3.onError(SingleAccountPublicClientApplication.java:214)
	at com.microsoft.identity.common.internal.controllers.CommandDispatcher$2.run(CommandDispatcher.java:177)
	at android.os.Handler.handleCallback(Handler.java:790)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at android.os.Looper.loop(Looper.java:164)
	at android.app.ActivityThread.main(ActivityThread.java:6494)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:807)
Caused by: com.microsoft.identity.common.exception.ClientException: Attempt to invoke virtual method 'long java.lang.Long.longValue()' on a null object reference
	at com.microsoft.identity.common.internal.controllers.ExceptionAdapter.baseExceptionFromException(ExceptionAdapter.java:252)
	at com.microsoft.identity.common.internal.controllers.CommandDispatcher.executeCommand(CommandDispatcher.java:143)
	at com.microsoft.identity.common.internal.controllers.CommandDispatcher.access$200(CommandDispatcher.java:53)
	at com.microsoft.identity.common.internal.controllers.CommandDispatcher$1.run(CommandDispatcher.java:95)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
	at java.lang.Thread.run(Thread.java:764)
Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'long java.lang.Long.longValue()' on a null object reference
	at com.microsoft.identity.common.internal.cache.MicrosoftStsAccountCredentialAdapter.getExpiresOn(MicrosoftStsAccountCredentialAdapter.java:262)
	at com.microsoft.identity.common.internal.cache.MicrosoftStsAccountCredentialAdapter.createAccessToken(MicrosoftStsAccountCredentialAdapter.java:73)
	at com.microsoft.identity.common.internal.cache.MicrosoftStsAccountCredentialAdapter.createAccessToken(MicrosoftStsAccountCredentialAdapter.java:47)
	at com.microsoft.identity.common.internal.cache.MsalOAuth2TokenCache.save(MsalOAuth2TokenCache.java:268)
	at com.microsoft.identity.common.internal.cache.MsalOAuth2TokenCache.saveAndLoadAggregatedAccountData(MsalOAuth2TokenCache.java:361)
	at com.microsoft.identity.common.internal.controllers.BaseController.renewAccessToken(BaseController.java:224)
	at com.microsoft.identity.client.internal.controllers.LocalMSALController.acquireTokenSilent(LocalMSALController.java:266)
	at com.microsoft.identity.common.internal.controllers.TokenCommand.execute(TokenCommand.java:73)
	at com.microsoft.identity.common.internal.controllers.TokenCommand.execute(TokenCommand.java:42)
	at com.microsoft.identity.common.internal.controllers.CommandDispatcher.executeCommand(CommandDispatcher.java:138)
```
As you can see, the `TokenResponse`'s `expires_in` value is not set due to the fact that there is also no `access_token` present in the response. This causes the error while trying to create the `AccessTokenRecord`.

**Comment by iambmelt** [comment]
Work item in `common` created -- this fix will be authored there and will be deployed in an upcoming release.

Currently working with internal team to determine release vehicle for fix; hotfix vs. monthly release

/cc @koesclem ^^

**Comment by iambmelt** [comment]
PR with proposed fix:
https://github.com/AzureAD/microsoft-authentication-library-for-android/pull/905

@koesclem Feel free to pull down this branch if you would like to try/test these proposed changes.

Branch info:
- MSAL: `iambmelt/common-793` ([link](https://github.com/AzureAD/microsoft-authentication-library-for-android/tree/iambmelt/common-793))
- Common: `iambmelt/793` ([link](https://github.com/AzureAD/microsoft-authentication-library-common-for-android/tree/iambmelt/793))

Clone info:
```
git clone --recurse-submodules -b iambmelt/common-793 https://github.com/AzureAD/microsoft-authentication-library-for-android.git
```

**Comment by iambmelt** [comment]
> PR with proposed fix:
> #905
> 
> @koesclem Feel free to pull down this branch if you would like to try/test these proposed changes.
> 
> Branch info:
> 
> * MSAL: `iambmelt/common-793` ([link](https://github.com/AzureAD/microsoft-authentication-library-for-android/tree/iambmelt/common-793))
> * Common: `iambmelt/793` ([link](https://github.com/AzureAD/microsoft-authentication-library-common-for-android/tree/iambmelt/793))
> 
> Clone info:
> 
> ```
> git clone --recurse-submodules -b iambmelt/common-793 https://github.com/AzureAD/microsoft-authentication-library-for-android.git
> ```

One more thing I'll add, I'm going to make additional changes to this PR before it merges to account for the behavior described [here](https://tools.ietf.org/html/rfc6749#section-3.3) in [RFC-6749](https://tools.ietf.org/html/rfc6749) which describes empty `scopes` in the response. This is not currently accounted for in these changes. I don't believe this will impact your scenario _directly_ on B2C as I don't think B2C makes use of this payload optimization. If this does end up impacting you, I should have that portion of these changes authored sometime tomorrow and available for test afterwards

**Comment by koesclem** [comment]
@iambmelt Using your mentioned fix #905 I was successfully able to fetch an `access_token` for `scope_of_resource1` and `scope_of_resource2` interactively (2 separate requests). And later perform silent token lookups from the cache for still valid (not expired) tokens for both scopes.

Issues still arise if those `access_token` in the cache are expired. Same, as not being in the cache at all this would trigger a token renewal using the `refresh_token`. As mentioned in [my previous comment](https://github.com/AzureAD/microsoft-authentication-library-for-android/issues/904#issuecomment-578687990) this leads to an error. I was able to find some documentation about the multi-resource-capability of the `refresh_token` in the `MsalOAuth2TokenCache`'s `removeRefreshTokensForAccount` method:
```
// AAD v1 & v2 support multi-resource refresh tokens, allowing us to use
// a single refresh token to service all of an account's requests.
// To ensure that only one refresh token is maintained for an account,
// refresh tokens are cleared from the cache for the account which is about to be
// saved (in the event that there was already a refresh token in the cache)
```
This maybe sheds some light on the intended behaviour.

Currently the `refresh_token` returned when fetching `scope_of_resource1` can later be used to silently renew the token for `scope_of_resource1`. It can **not** be used to silently renew the token for `scope_of_resource2`.

**Comment by iambmelt** [comment]
> @iambmelt Using your mentioned fix #905 I was successfully able to fetch an `access_token` for `scope_of_resource1` and `scope_of_resource2` interactively (2 separate requests). And later perform silent token lookups from the cache for still valid (not expired) tokens for both scopes.
> 
> Issues still arise if those `access_token` in the cache are expired. Same, as not being in the cache at all this would trigger a token renewal using the `refresh_token`. As mentioned in [my previous comment](https://github.com/AzureAD/microsoft-authentication-library-for-android/issues/904#issuecomment-578687990) this leads to an error. I was able to find some documentation about the multi-resource-capability of the `refresh_token` in the `MsalOAuth2TokenCache`'s `removeRefreshTokensForAccount` method:
> 
> ```
> // AAD v1 & v2 support multi-resource refresh tokens, allowing us to use
> // a single refresh token to service all of an account's requests.
> // To ensure that only one refresh token is maintained for an account,
> // refresh tokens are cleared from the cache for the account which is about to be
> // saved (in the event that there was already a refresh token in the cache)
> ```
> 
> This maybe sheds some light on the intended behaviour.
> 
> Currently the `refresh_token` returned when fetching `scope_of_resource1` can later be used to silently renew the token for `scope_of_resource1`. It can **not** be used to silently renew the token for `scope_of_resource2`.

Thanks @koesclem for confirming the behavior you observed with the change in #905 -- I am still following up with our B2C backend team and will provide an update on the expected RT behavior once I hear back

**Comment by solidstore** [comment]
@iambmelt We are also experiencing similar problems obtaining an access token for a second resource using the MSAL. 

In our case we are using a ConfidentialClient. 

Based on a very old post [here](https://www.cloudidentity.com/blog/2013/10/14/adal-windows-azure-ad-and-multi-resource-refresh-tokens/) we have tried adding the "resource" parameter.

`string accessToken = null;
var scope = new string[] { Globals.WriteTasksScope_Api2 };

IConfidentialClientApplication cca = MsalAppBuilder.BuildConfidentialClientApplication();
var accounts = await cca.GetAccountsAsync();
AuthenticationResult result = await cca.AcquireTokenSilent(scope, accounts.FirstOrDefault())
	.WithExtraQueryParameters(new Dictionary<string, string>
	{
		{ "resource", "https://myb2c.onmicrosoft.com/api2" }
	})
	.ExecuteAsync();

accessToken = result.AccessToken; // null!`

We do not receive an error but the result does not contain an AccessToken

**Comment by iambmelt** [comment]
@koesclem 

I received the following guidance from our B2C team -- their suggestion is to configure the validation on the RP to check for permissions embedded in the `scp` claim, of which multiple can be applied
>So b2c scopes are as follows:
https://tenantname.onmicrosoft.com/resourceName/permissionName
A single user flow maps to a single resource / access token by design, however it can have multiple permissions. <br/><br/>The customer should model their apps as permissions of a single resource. 
Ie. https://tenantname.onmicrosoft.com/resourceName/app1, https://tenantname.onmicrosoft.com/resourceName/app2
Then they will get a single access token with a single audience, but multiple approved permissions/scopes in the `scp` claim.
During token validation they must validate that the scopes in the token have permission to access the app.

**Comment by koesclem** [comment]
@iambmelt Thanks for this follow-up!

Just to make sure and because the term "app" is nowadays used excessively - when reading the above statement and the term "app" is mentioned, we are **not** talking about the Mobile-App but e.g. API-Services the Mobile-App uses to access those resources - correct?

Also to summarize: In a B2C Mobile Context, working with multiple resources is **not** supported by the MSAL. It is recommended to work with a **single** resource while using app-based permissions checks in the scope claim?

If those statements are correct, I would suggest to integrate the results of our conversation in the existing B2C MSAL documentation as it seems to have huge impacts on making architectural decisions when designing services intended to be used in a Mobile context. Also integration guides of the MSAL and the MSAL itself would imo suggest that using multiple resources is indeed supported (e.g. the aforementioned method `withOtherScopesToAuthorize()`).

**Comment by iambmelt** [comment]
> @iambmelt Thanks for this follow-up!
> 
> Just to make sure and because the term "app" is nowadays used excessively - when reading the above statement and the term "app" is mentioned, we are **not** talking about the Mobile-App but e.g. API-Services the Mobile-App uses to access those resources - correct?

That's correct.

> Also to summarize: In a B2C Mobile Context, working with multiple resources is **not** supported by the MSAL. It is recommended to work with a **single** resource while using app-based permissions checks in the scope claim?

Working with multiple resources associated to a single _policy_ (journey/flow) is not supported. Instead, the developer is expected to associate multiple scopes (permissions) to a single policy and use the embedded `scp` claim to validate the token on the middleware

> If those statements are correct, I would suggest to integrate the results of our conversation in the existing B2C MSAL documentation as it seems to have huge impacts on making architectural decisions when designing services intended to be used in a Mobile context. Also integration guides of the MSAL and the MSAL itself would imo suggest that using multiple resources is indeed supported (e.g. the aforementioned method `withOtherScopesToAuthorize()`).

Thanks for this feedback -- as this experience represents a departure from our typical AAD experience, we'll have to work with our PM team to iron out how this documentation is surfaced. /cc @hamiltonha @valnav

**Comment by iambmelt** [comment]
Closing as guidance is provided in https://github.com/AzureAD/microsoft-authentication-library-for-android/issues/904#issuecomment-581687190
